From bc2d58b852664d9aae648ea78b923c4659b05258 Mon Sep 17 00:00:00 2001
From: Hao Chen <chenhao@loongson.cn>
Date: Tue, 11 Jun 2024 11:18:14 +0800
Subject: [PATCH 07/15] FROMLIST: LoongArch64: Optimize functions in dct.S file
 - Patch 2

The optimized functions are as follows:
quant_lsx/quant_lasx
dequant_normal_lsx/dequant_normal_lasx
dequant_scaling_lsx/dequant_scaling_lasx
nquant_lsx/nquant_lasx

Link: https://bitbucket.org/multicoreware/x265_git/pull-requests/31
Signed-off-by: Mingcong Bai <jeffbai@aosc.io>
---
 source/common/loongarch64/asm-primitives.cpp |   8 +
 source/common/loongarch64/dct.S              | 480 +++++++++++++++++++
 source/common/loongarch64/dct.h              |   8 +
 3 files changed, 496 insertions(+)

diff --git a/source/common/loongarch64/asm-primitives.cpp b/source/common/loongarch64/asm-primitives.cpp
index ff1e7b0a5..b79297367 100644
--- a/source/common/loongarch64/asm-primitives.cpp
+++ b/source/common/loongarch64/asm-primitives.cpp
@@ -215,6 +215,10 @@ void setupAssemblyPrimitives(EncoderPrimitives &p, int cpuMask)
         p.idst4x4              = PFX(idst4_lsx);
         p.scanPosLast          = PFX(scanPosLast_lsx);
         p.costCoeffNxN         = PFX(costCoeffNxN_lsx);
+        p.quant                = PFX(quant_lsx);
+        p.dequant_normal       = PFX(dequant_normal_lsx);
+        p.dequant_scaling      = PFX(dequant_scaling_lsx);
+        p.nquant               = PFX(nquant_lsx);
     }
 #endif /* HAVE_LSX */
 
@@ -288,6 +292,10 @@ void setupAssemblyPrimitives(EncoderPrimitives &p, int cpuMask)
         p.cu[BLOCK_32x32].dct  = PFX(dct32_lasx);
         p.cu[BLOCK_16x16].idct = PFX(idct16_lasx);
         p.cu[BLOCK_32x32].idct = PFX(idct32_lasx);
+        p.quant                = PFX(quant_lasx);
+        p.dequant_normal       = PFX(dequant_normal_lasx);
+        p.dequant_scaling      = PFX(dequant_scaling_lasx);
+        p.nquant               = PFX(nquant_lasx);
     }
 #endif /* HAVE_LASX */
 }
diff --git a/source/common/loongarch64/dct.S b/source/common/loongarch64/dct.S
index c9d25daf4..7f9b05b42 100644
--- a/source/common/loongarch64/dct.S
+++ b/source/common/loongarch64/dct.S
@@ -5651,3 +5651,483 @@ function x265_costCoeffNxN_lsx
 5:
     and           a0,    t8,   t7
 endfunc
+
+/* uint32_t quant_c(const int16_t* coef, const int32_t* quantCoeff,
+   int32_t* deltaU, int16_t* qCoef, int qBits, int add, int numCoeff)
+*/
+function x265_quant_lsx
+    srli.d        t0,    a6,   4     //numCoeff / 16
+    vreplgr2vr.w  vr20,  a4         //qBits
+    vreplgr2vr.w  vr22,  a5         //add
+    vldi          vr23,  0          //numsig
+    vldi          vr10,  0
+    vldi          vr11,  0x401
+    vsubi.wu      vr21,  vr20, 8    //qBits8
+1:
+    vld           vr0,   a0,   0     //level
+    vld           vr1,   a0,   16
+    vld           vr2,   a1,   0     //quantCoeff
+    vld           vr3,   a1,   16
+    vld           vr4,   a1,   32
+    vld           vr5,   a1,   48
+    vsigncov.h    vr6,   vr0,  vr11
+    vsigncov.h    vr7,   vr1,  vr11
+    vadda.h       vr8,   vr0,  vr10  //abs(level)
+    vadda.h       vr9,   vr1,  vr10
+    vor.v         vr6,   vr6,  vr11  //sign
+    vor.v         vr7,   vr7,  vr11
+    vilvl.h       vr12,  vr10, vr8
+    vilvh.h       vr13,  vr10, vr8
+    vilvl.h       vr14,  vr10, vr9
+    vilvh.h       vr15,  vr10, vr9
+    vmul.w        vr12,  vr12, vr2  //tmplevel
+    vmul.w        vr13,  vr13, vr3
+    vmul.w        vr14,  vr14, vr4
+    vmul.w        vr15,  vr15, vr5
+    vadd.w        vr2,   vr12, vr22   //templevel + add
+    vadd.w        vr3,   vr13, vr22
+    vadd.w        vr4,   vr14, vr22
+    vadd.w        vr5,   vr15, vr22
+    vsra.w        vr2,   vr2,  vr20   //level
+    vsra.w        vr3,   vr3,  vr20
+    vsra.w        vr4,   vr4,  vr20
+    vsra.w        vr5,   vr5,  vr20
+    vsll.w        vr16,  vr2,  vr20   //level << qBits
+    vsll.w        vr17,  vr3,  vr20
+    vsll.w        vr18,  vr4,  vr20
+    vsll.w        vr19,  vr5,  vr20
+    vsub.w        vr12,  vr12, vr16
+    vsub.w        vr13,  vr13, vr17
+    vsub.w        vr14,  vr14, vr18
+    vsub.w        vr15,  vr15, vr19
+    vsra.w        vr12,  vr12, vr21
+    vsra.w        vr13,  vr13, vr21
+    vsra.w        vr14,  vr14, vr21
+    vsra.w        vr15,  vr15, vr21
+    vst           vr12,  a2,   0
+    vst           vr13,  a2,   16
+    vst           vr14,  a2,   32
+    vst           vr15,  a2,   48
+    vseq.w        vr16,  vr2,  vr10
+    vseq.w        vr17,  vr3,  vr10
+    vseq.w        vr18,  vr4,  vr10
+    vseq.w        vr19,  vr5,  vr10
+    vpickev.h     vr8,   vr17, vr16
+    vpickev.h     vr9,   vr19, vr18
+    vilvl.h       vr16,  vr6,  vr6
+    vilvl.h       vr18,  vr7,  vr7
+    vandn.v       vr12,  vr8,  vr11
+    vandn.v       vr13,  vr9,  vr11
+    vilvh.h       vr17,  vr6,  vr6
+    vilvh.h       vr19,  vr7,  vr7
+    vsigncov.w    vr0,   vr16, vr2
+    vsigncov.w    vr8,   vr18, vr4
+    vsigncov.w    vr1,   vr17, vr3
+    vsigncov.w    vr9,   vr19, vr5
+    vadd.h        vr23,  vr23, vr12
+    vssrani.h.w   vr1,   vr0,  0
+    vssrani.h.w   vr9,   vr8,  0
+    vadd.h        vr23,  vr23, vr13
+    vst           vr1,   a3,   0
+    vst           vr9,   a3,   16
+    addi.d        t0,    t0,   -1
+    addi.d        a0,    a0,   32
+    addi.d        a1,    a1,   64
+    addi.d        a2,    a2,   64
+    addi.d        a3,    a3,   32
+    blt           zero,  t0,   1b
+    vhaddw.wu.hu  vr0,   vr23, vr23
+    vhaddw.du.wu  vr1,   vr0,  vr0
+    vhaddw.qu.du  vr2,   vr1,  vr1
+    vpickve2gr.w  a0,    vr2,  0
+endfunc
+
+function x265_quant_lasx
+    srli.d         t0,    a6,   4     //numCoeff / 16
+    xvreplgr2vr.w  xr20,  a4         //qBits
+    xvreplgr2vr.w  xr22,  a5         //add
+    xvldi          xr23,  0          //numsig
+    xvldi          xr10,  0
+    xvldi          xr11,  0x401
+    xvsubi.wu      xr21,  xr20, 8    //qBits8
+2:
+    xvld           xr0,   a0,   0     //level
+    xvld           xr1,   a1,   0     //quantCoeff
+    xvld           xr2,   a1,   32
+    xvpermi.d      xr0,   xr0,  0xD8
+    xvsigncov.h    xr6,   xr0,  xr11
+    xvadda.h       xr8,   xr0,  xr10  //abs(level)
+    xvor.v         xr6,   xr6,  xr11  //sign
+    xvilvl.h       xr12,  xr10, xr8
+    xvilvh.h       xr13,  xr10, xr8
+    xvmul.w        xr12,  xr12, xr1  //tmplevel
+    xvmul.w        xr13,  xr13, xr2
+    xvadd.w        xr2,   xr12, xr22   //templevel + add
+    xvadd.w        xr3,   xr13, xr22
+    xvsra.w        xr2,   xr2,  xr20   //level
+    xvsra.w        xr3,   xr3,  xr20
+    xvsll.w        xr16,  xr2,  xr20   //level << qBits
+    xvsll.w        xr17,  xr3,  xr20
+    xvsub.w        xr12,  xr12, xr16
+    xvsub.w        xr13,  xr13, xr17
+    xvsra.w        xr12,  xr12, xr21
+    xvsra.w        xr13,  xr13, xr21
+    xvseq.w        xr16,  xr2,  xr10
+    xvseq.w        xr17,  xr3,  xr10
+    xvst           xr12,  a2,   0
+    xvst           xr13,  a2,   32
+    xvpickev.h     xr8,   xr17, xr16
+    xvilvl.h       xr16,  xr6,  xr6
+    xvilvh.h       xr17,  xr6,  xr6
+    xvandn.v       xr12,  xr8,  xr11
+    xvsigncov.w    xr0,   xr16, xr2
+    xvsigncov.w    xr1,   xr17, xr3
+    xvadd.h        xr23,  xr23, xr12
+    xvssrani.h.w   xr1,   xr0,  0
+    addi.d         t0,    t0,   -1
+    addi.d         a0,    a0,   32
+    xvpermi.d      xr2,   xr1,  0xD8
+    addi.d         a1,    a1,   64
+    addi.d         a2,    a2,   64
+    xvst           xr2,   a3,   0
+    addi.d         a3,    a3,   32
+    blt            zero,  t0,   2b
+    xvhaddw.wu.hu  xr0,   xr23, xr23
+    xvhaddw.du.wu  xr1,   xr0,  xr0
+    xvhaddw.qu.du  xr2,   xr1,  xr1
+    xvpickve2gr.w  t0,    xr2,  0
+    xvpickve2gr.w  t1,    xr2,  4
+    add.d          a0,    t0,   t1
+endfunc
+
+/* void dequant_normal(const int16_t* quantCoef, int16_t* coef, int num,
+   int scale, int shift)
+*/
+function x265_dequant_normal_lsx
+    srli.d        t0,    a2,   4     //num / 16
+    andi          t1,    a2,   15
+    vreplgr2vr.w  vr20,  a3         //scale
+    vreplgr2vr.w  vr21,  a4
+1:
+    beqz          t0,    2f
+    vld           vr0,   a0,   0     //quantCoef
+    vld           vr1,   a0,   16
+    addi.d        a0,    a0,   32
+    vsllwil.w.h   vr2,   vr0,  0
+    vsllwil.w.h   vr4,   vr1,  0
+    vexth.w.h     vr3,   vr0
+    vexth.w.h     vr5,   vr1
+    vmul.w        vr6,   vr2,  vr20
+    vmul.w        vr7,   vr3,  vr20
+    vmul.w        vr8,   vr4,  vr20
+    vmul.w        vr9,   vr5,  vr20
+    vssrarn.h.w   vr0,   vr6,  vr21
+    vssrarn.h.w   vr1,   vr7,  vr21
+    vssrarn.h.w   vr2,   vr8,  vr21
+    vssrarn.h.w   vr3,   vr9,  vr21
+    vilvl.d       vr7,   vr1,  vr0
+    vilvl.d       vr9,   vr3,  vr2
+    addi.d        t0,    t0,   -1
+    vst           vr7,   a1,   0
+    vst           vr9,   a1,   16
+    addi.d        a1,    a1,   32
+    b             1b
+2:
+    beqz          t1,    3f
+    vld           vr0,   a0,   0
+    vsllwil.w.h   vr1,   vr0,  0
+    vexth.w.h     vr2,   vr0
+    vmul.w        vr3,   vr1,  vr20
+    vmul.w        vr4,   vr2,  vr20
+    vssrarn.h.w   vr5,   vr3,  vr21
+    vssrarn.h.w   vr6,   vr4,  vr21
+    vilvl.d       vr7,   vr6,  vr5
+    vst           vr7,   a1,   0
+3:
+endfunc
+
+function x265_dequant_normal_lasx
+    srli.d         t0,    a2,   4     //num / 16
+    andi           t1,    a2,   15
+    xvreplgr2vr.w  xr20,  a3         //scale
+    xvreplgr2vr.w  xr21,  a4
+1:
+    beqz           t0,    2f
+    xvld           xr0,   a0,   0     //quantCoef
+    addi.d         a0,    a0,   32
+    xvsllwil.w.h   xr1,   xr0,  0
+    xvexth.w.h     xr2,   xr0
+    xvmul.w        xr3,   xr1,  xr20
+    xvmul.w        xr4,   xr2,  xr20
+    addi.d         t0,    t0,   -1
+    xvssrarn.h.w   xr5,   xr3,  xr21
+    xvssrarn.h.w   xr6,   xr4,  xr21
+    xvilvl.d       xr7,   xr6,  xr5
+    xvst           xr7,   a1,   0
+    addi.d         a1,    a1,   32
+    blt            zero,  t0,   1b
+2:
+    beqz           t1,    3f
+    vld            vr0,   a0,   0
+    vext2xv.w.h    xr1,   xr0
+    xvmul.w        xr2,   xr1,  xr20
+    xvssrarn.h.w   xr3,   xr2,  xr21
+    xvpermi.d      xr4,   xr3,  0xD8
+    vst            vr4,   a1,   0
+3:
+endfunc
+
+/* void dequant_scaling(const int16_t* quantCoef, const int32_t* deQuantCoef,
+   int16_t* coef, int num, int per, int shift)
+*/
+function x265_dequant_scaling_lsx
+    addi.d         a5,    a5,   4
+    srli.d         t0,    a3,   4
+    bge            a4,    a5,   1f
+    sub.d          t3,    a5,   a4
+    vreplgr2vr.w   vr20,  t3
+2:
+    beqz           t0,    3f
+    vld            vr0,   a0,   0
+    vld            vr1,   a0,   16
+    vld            vr2,   a1,   0
+    vld            vr3,   a1,   16
+    vld            vr4,   a1,   32
+    vld            vr5,   a1,   48
+    vsllwil.w.h    vr6,   vr0,  0
+    vsllwil.w.h    vr8,   vr1,  0
+    vexth.w.h      vr7,   vr0
+    vexth.w.h      vr9,   vr1
+    vmul.w         vr10,  vr6,  vr2
+    vmul.w         vr11,  vr7,  vr3
+    vmul.w         vr12,  vr8,  vr4
+    vmul.w         vr13,  vr9,  vr5
+    vssrarn.h.w    vr0,   vr10, vr20
+    vssrarn.h.w    vr1,   vr11, vr20
+    vssrarn.h.w    vr2,   vr12, vr20
+    vssrarn.h.w    vr3,   vr13, vr20
+    addi.d         t0,    t0,   -1
+    vilvl.d        vr4,   vr1,  vr0
+    vilvl.d        vr5,   vr3,  vr2
+    addi.d         a0,    a0,   32
+    addi.d         a1,    a1,   64
+    vst            vr4,   a2,   0
+    vst            vr5,   a2,   16
+    addi.d         a2,    a2,   32
+    b              2b
+1:
+    sub.d          t3,    a4,   a5
+    vreplgr2vr.w   vr20,  t3
+4:
+    beqz           t0,    3f
+    vld            vr0,   a0,   0
+    vld            vr1,   a0,   16
+    vld            vr2,   a1,   0
+    vld            vr3,   a1,   16
+    vld            vr4,   a1,   32
+    vld            vr5,   a1,   48
+    vsllwil.w.h    vr6,   vr0,  0
+    vsllwil.w.h    vr8,   vr1,  0
+    vexth.w.h      vr7,   vr0
+    vexth.w.h      vr9,   vr1
+    vmul.w         vr10,  vr6,  vr2
+    vmul.w         vr11,  vr7,  vr3
+    vmul.w         vr12,  vr8,  vr4
+    vmul.w         vr13,  vr9,  vr5
+    vsat.w         vr0,   vr10, 31
+    vsat.w         vr1,   vr11, 31
+    vsat.w         vr2,   vr12, 31
+    vsat.w         vr3,   vr13, 31
+    vsll.w         vr4,   vr0,  vr20
+    vsll.w         vr5,   vr1,  vr20
+    vsll.w         vr6,   vr2,  vr20
+    vsll.w         vr7,   vr3,  vr20
+    addi.d         a0,    a0,   32
+    vssrani.h.w    vr5,   vr4,  0
+    vssrani.h.w    vr7,   vr6,  0
+    addi.d         a1,    a1,   64
+    addi.d         t0,    t0,   -1
+    vst            vr5,   a2,   0
+    vst            vr7,   a2,   16
+    addi.d         a2,    a2,   32
+    b              4b
+3:
+endfunc
+
+function x265_dequant_scaling_lasx
+    addi.d         a5,    a5,   4
+    srli.d         t0,    a3,   4
+    bge            a4,    a5,   1f
+    sub.d          t3,    a5,   a4
+    xvreplgr2vr.w  xr20,  t3
+2:
+    beqz           t0,    4f
+    xvld           xr0,   a0,   0
+    xvld           xr2,   a1,   0
+    xvld           xr3,   a1,   32
+    xvpermi.d      xr1,   xr0,  0xD8
+    xvsllwil.w.h   xr6,   xr1,  0
+    xvexth.w.h     xr7,   xr1
+    xvmul.w        xr10,  xr6,  xr2
+    xvmul.w        xr11,  xr7,  xr3
+    xvssrarn.h.w   xr0,   xr10, xr20
+    xvssrarn.h.w   xr1,   xr11, xr20
+    addi.d         t0,    t0,   -1
+    xvilvl.d       xr4,   xr1,  xr0
+    addi.d         a0,    a0,   32
+    xvpermi.d      xr5,   xr4,  0xD8
+    addi.d         a1,    a1,   64
+    xvst           xr5,   a2,   0
+    addi.d         a2,    a2,   32
+    b              2b
+1:
+    sub.d          t3,    a4,   a5
+    xvreplgr2vr.w  xr20,  t3
+3:
+    beqz           t0,    4f
+    xvld           xr0,   a0,   0
+    xvld           xr2,   a1,   0
+    xvld           xr3,   a1,   32
+    xvpermi.d      xr1,   xr0,  0xD8
+    xvsllwil.w.h   xr6,   xr1,  0
+    xvexth.w.h     xr7,   xr1
+    xvmul.w        xr10,  xr6,  xr2
+    xvmul.w        xr11,  xr7,  xr3
+    xvsat.w        xr0,   xr10, 31
+    xvsat.w        xr1,   xr11, 31
+    xvsll.w        xr4,   xr0,  xr20
+    xvsll.w        xr5,   xr1,  xr20
+    addi.d         a0,    a0,   32
+    xvssrani.h.w   xr5,   xr4,  0
+    addi.d         a1,    a1,   64
+    xvpermi.d      xr6,   xr5,  0xD8
+    addi.d         t0,    t0,   -1
+    xvst           xr6,   a2,   0
+    addi.d         a2,    a2,   32
+    b              3b
+4:
+endfunc
+
+/* uint32_t nquant_c(const int16_t* coef, const int32_t* quantCoeff,
+   int16_t* qCoef, int qBits, int add, int numCoeff)
+*/
+function x265_nquant_lsx
+    srli.d        t0,    a5,   4     //numCoeff / 16
+    vreplgr2vr.w  vr20,  a3         //qBits
+    vreplgr2vr.w  vr22,  a4         //add
+    vldi          vr23,  0          //numsig
+    vldi          vr10,  0
+    vldi          vr11,  0x401
+1:
+    vld           vr0,   a0,   0     //level
+    vld           vr1,   a0,   16
+    vld           vr2,   a1,   0     //quantCoeff
+    vld           vr3,   a1,   16
+    vld           vr4,   a1,   32
+    vld           vr5,   a1,   48
+    vsigncov.h    vr6,   vr0,  vr11
+    vsigncov.h    vr7,   vr1,  vr11
+    vadda.h       vr8,   vr0,  vr10  //abs(level)
+    vadda.h       vr9,   vr1,  vr10
+    vor.v         vr6,   vr6,  vr11  //sign
+    vor.v         vr7,   vr7,  vr11
+    vilvl.h       vr12,  vr10, vr8
+    vilvh.h       vr13,  vr10, vr8
+    vilvl.h       vr14,  vr10, vr9
+    vilvh.h       vr15,  vr10, vr9
+    vmul.w        vr12,  vr12, vr2  //tmplevel
+    vmul.w        vr13,  vr13, vr3
+    vmul.w        vr14,  vr14, vr4
+    vmul.w        vr15,  vr15, vr5
+    vadd.w        vr2,   vr12, vr22   //templevel + add
+    vadd.w        vr3,   vr13, vr22
+    vadd.w        vr4,   vr14, vr22
+    vadd.w        vr5,   vr15, vr22
+    vsra.w        vr2,   vr2,  vr20   //level
+    vsra.w        vr3,   vr3,  vr20
+    vsra.w        vr4,   vr4,  vr20
+    vsra.w        vr5,   vr5,  vr20
+
+    vseq.w        vr16,  vr2,  vr10
+    vseq.w        vr17,  vr3,  vr10
+    vseq.w        vr18,  vr4,  vr10
+    vseq.w        vr19,  vr5,  vr10
+    vpickev.h     vr8,   vr17, vr16
+    vpickev.h     vr9,   vr19, vr18
+    vilvl.h       vr16,  vr6,  vr6
+    vilvl.h       vr18,  vr7,  vr7
+    vandn.v       vr12,  vr8,  vr11
+    vandn.v       vr13,  vr9,  vr11
+    vilvh.h       vr17,  vr6,  vr6
+    vilvh.h       vr19,  vr7,  vr7
+    vsigncov.w    vr0,   vr16, vr2
+    vsigncov.w    vr8,   vr18, vr4
+    vsigncov.w    vr1,   vr17, vr3
+    vsigncov.w    vr9,   vr19, vr5
+    vadd.h        vr23,  vr23, vr12
+    vssrani.h.w   vr1,   vr0,  0
+    vssrani.h.w   vr9,   vr8,  0
+    vadd.h        vr23,  vr23, vr13
+    vadda.h       vr1,   vr1,  vr10
+    vadda.h       vr9,   vr9,  vr10
+    vst           vr1,   a2,   0
+    vst           vr9,   a2,   16
+    addi.d        t0,    t0,   -1
+    addi.d        a0,    a0,   32
+    addi.d        a1,    a1,   64
+    addi.d        a2,    a2,   32
+    blt           zero,  t0,   1b
+    vhaddw.wu.hu  vr0,   vr23, vr23
+    vhaddw.du.wu  vr1,   vr0,  vr0
+    vhaddw.qu.du  vr2,   vr1,  vr1
+    vpickve2gr.w  a0,    vr2,  0
+endfunc
+
+function x265_nquant_lasx
+    srli.d         t0,    a5,   4     //numCoeff / 16
+    xvreplgr2vr.w  xr20,  a3         //qBits
+    xvreplgr2vr.w  xr22,  a4         //add
+    xvldi          xr23,  0          //numsig
+    xvldi          xr10,  0
+    xvldi          xr11,  0x401
+1:
+    xvld           xr0,   a0,   0     //level
+    xvld           xr1,   a1,   0     //quantCoeff
+    xvld           xr2,   a1,   32
+    xvpermi.d      xr0,   xr0,  0xD8
+    xvsigncov.h    xr6,   xr0,  xr11
+    xvadda.h       xr8,   xr0,  xr10  //abs(level)
+    xvor.v         xr6,   xr6,  xr11  //sign
+    xvilvl.h       xr12,  xr10, xr8
+    xvilvh.h       xr13,  xr10, xr8
+    xvmul.w        xr12,  xr12, xr1  //tmplevel
+    xvmul.w        xr13,  xr13, xr2
+    xvadd.w        xr2,   xr12, xr22   //templevel + add
+    xvadd.w        xr3,   xr13, xr22
+    xvsra.w        xr2,   xr2,  xr20   //level
+    xvsra.w        xr3,   xr3,  xr20
+    xvseq.w        xr16,  xr2,  xr10
+    xvseq.w        xr17,  xr3,  xr10
+    xvpickev.h     xr8,   xr17, xr16
+    xvilvl.h       xr16,  xr6,  xr6
+    xvilvh.h       xr17,  xr6,  xr6
+    xvandn.v       xr12,  xr8,  xr11
+    xvsigncov.w    xr0,   xr16, xr2
+    xvsigncov.w    xr1,   xr17, xr3
+    xvadd.h        xr23,  xr23, xr12
+    xvssrani.h.w   xr1,   xr0,  0
+    addi.d         t0,    t0,   -1
+    addi.d         a0,    a0,   32
+    xvpermi.d      xr2,   xr1,  0xD8
+    addi.d         a1,    a1,   64
+    xvadda.h       xr2,   xr2,  xr10
+    xvst           xr2,   a2,   0
+    addi.d         a2,    a2,   32
+    blt            zero,  t0,   1b
+    xvhaddw.wu.hu  xr0,   xr23, xr23
+    xvhaddw.du.wu  xr1,   xr0,  xr0
+    xvhaddw.qu.du  xr2,   xr1,  xr1
+    xvpickve2gr.w  t0,    xr2,  0
+    xvpickve2gr.w  t1,    xr2,  4
+    add.d          a0,    t0,   t1
+endfunc
diff --git a/source/common/loongarch64/dct.h b/source/common/loongarch64/dct.h
index 4ea8d5f20..f78422e61 100644
--- a/source/common/loongarch64/dct.h
+++ b/source/common/loongarch64/dct.h
@@ -44,11 +44,19 @@ int PFX(scanPosLast_lsx)(const uint16_t *scan, const coeff_t *coeff,
 uint32_t PFX(costCoeffNxN_lsx)(const uint16_t *scan, const coeff_t *coeff, intptr_t trSize,
                                uint16_t *absCoeff, const uint8_t *tabSigCtx, uint32_t scanFlagMask,
                                uint8_t *baseCtx, int offset, int scanPosSigOff, int subPosBase);
+uint32_t PFX(quant_lsx)(const int16_t* coef, const int32_t* quantCoeff, int32_t* deltaU, int16_t* qCoef, int qBits, int add, int numCoeff);
+void PFX(dequant_normal_lsx)(const int16_t* quantCoef, int16_t* coef, int num, int scale, int shift);
+void PFX(dequant_scaling_lsx)(const int16_t* quantCoef, const int32_t* deQuantCoef, int16_t* coef, int num, int per, int shift);
+uint32_t PFX(nquant_lsx)(const int16_t* coef, const int32_t* quantCoeff, int16_t* qCoef, int qBits, int add, int numCoeff);
 
 void PFX(dct16_lasx)(const int16_t* src, int16_t* dst, intptr_t srcStride);
 void PFX(dct32_lasx)(const int16_t* src, int16_t* dst, intptr_t srcStride);
 void PFX(idct16_lasx)(const int16_t* src, int16_t* dst, intptr_t dstStride);
 void PFX(idct32_lasx)(const int16_t* src, int16_t* dst, intptr_t dstStride);
+uint32_t PFX(quant_lasx)(const int16_t* coef, const int32_t* quantCoeff, int32_t* deltaU, int16_t* qCoef, int qBits, int add, int numCoeff);
+void PFX(dequant_normal_lasx)(const int16_t* quantCoef, int16_t* coef, int num, int scale, int shift);
+void PFX(dequant_scaling_lasx)(const int16_t* quantCoef, const int32_t* deQuantCoef, int16_t* coef, int num, int per, int shift);
+uint32_t PFX(nquant_lasx)(const int16_t* coef, const int32_t* quantCoeff, int16_t* qCoef, int qBits, int add, int numCoeff);
 
 #ifdef __cplusplus
 }
-- 
2.52.0

